### PR TITLE
[SPARK-7939][SQL] Add conf to enable/disable partition column type inference

### PR DESCRIPTION
**Pull request by viirya**
JIRA: https://issues.apache.org/jira/browse/SPARK-7939

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala [hunk 1]
```diff
@@ -71,6 +71,9 @@ private[spark] object SQLConf {
   // Whether to perform partition discovery when loading external data sources.  Default to true.
   val PARTITION_DISCOVERY_ENABLED = "spark.sql.sources.partitionDiscovery.enabled"
 
+  // Whether to perform partition column type inference. Default to false.
+  val PARTITION_COLUMN_TYPE_INFERENCE = "spark.sql.sources.partitionColumnTypeInference.enabled"
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala]: We probably want to set this to `true` to keep the default behavior backwards compatible with Spark 1.3.x.

**Comment by liancheng** [comment]
This LGTM except for the default value of the added configuration. Also, would you mind to add this to the SQL programming guide? Thanks!

**Comment by liancheng** [comment]
Thanks for fixing this! I'm merging this to master.

Another issue that needn't to be solved in this PR I'd like to mention is, it would be better to make the flag a data source relation option rather than a `SQLConf` option. However, currently `BaseRelation` doesn't expose data sources options in its interface, thus `HadoopFsRelation` doesn't have access to data source options. We may add a `HadoopFsRelation.options` method in a future PR and move the new flag as a data source option.

**Comment by viirya** [comment]
@liancheng Thanks for explanation.

**Pull request closed without merging**